### PR TITLE
tests:Add new AMD and Intel templates to wrmsr/cpuid snapshot/restore tests

### PR DESCRIPTION
**Pull request by sudanl0**
## Changes

- Use msr_cpu_template fixtures to add the new AMD (T2A) and Intel (T2CL) templates.
- Updating the MSR 0x277 in the wrmsr snapshot test resulted in a vm crash so remove it from the wrmsr_list.txt to make the test common for AMD and Intel
- Add a delay of 250ms before taking a snapshot to avoid a crash seen during restore on T2A.

## Reason

- Currently the wrmsr and cpuid snapshot/restore only supports Intel T2S template. It needs to be extended to support the new AMD and Intel templates.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

~~- [ ] If a specific issue led to this PR, this PR closes the issue.~~
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
~~- [ ] API changes follow the [Runbook for Firecracker API changes][2].~~
~~- [ ] User-facing changes are mentioned in `CHANGELOG.md`.~~
- [x] All added/changed functionality is tested.
- [x] New `TODO`s link to an issue.
- [x] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

~~- [ ] This functionality can be added in [`rust-vmm`][1].~~

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: tests/integration_tests/functional/test_cpu_features.py [hunk 1]
```diff
@@ -729,6 +717,7 @@ def _test_cpu_cpuid_snapshot(context):
         / context.kernel.base_name()
         / cpu_template_dir
     )
+    shutil.rmtree(snapshot_artifacts_dir, ignore_errors=True)
```
Comment (kalyazin) [inline tests/integration_tests/functional/test_cpu_features.py:720]: This can go in a separate commit as it is an independent fix, as I understand?
We can then use conventional tags in those commits: `fix(tests): ` and `feature(tests): `.
Also, can be more specific in the commit message on which templates are being tested.

File: tests/integration_tests/functional/test_cpu_features.py [hunk 2]
```diff
@@ -428,6 +428,10 @@ def _test_cpu_wrmsr_snapshot(context):
     msrs_before_fname = Path(snapshot_artifacts_dir) / shared_names["msrs_before_fname"]
 
     dump_msr_state_to_file(msrs_before_fname, vm.ssh, shared_names)
+    # On T2A, the restore test fails with error "cannot allocate memory" so,
+    # adding delay below as a workaround to unblock the tests for now.
+    # TODO: Debug the issue and remove this delay.
```
Comment (kalyazin) [inline tests/integration_tests/functional/test_cpu_features.py:433]: Looks like we decided some time back that all new TODOs in the code need a reference to a related Github issue.

File: resources/tests/msr/wrmsr_list.txt [hunk 3]
```diff
@@ -14,7 +14,6 @@
 0x26d 0x1
 0x26e 0x1
 0x26f 0x1
-0x277 0x7040600070407
```
Comment (kalyazin) [inline resources/tests/msr/wrmsr_list.txt:17]: It's better to mention this is in the commit message as well.

**Review by kalyazin — DISMISSED** [review]

File: tests/integration_tests/functional/test_cpu_features.py [hunk 4]
```diff
@@ -452,21 +457,13 @@ def _test_cpu_wrmsr_snapshot(context):
...
-)
 @pytest.mark.skipif(
     utils.get_kernel_version(level=1) not in SUPPORTED_KERNELS,
     reason=f"Supported kernels are {SUPPORTED_KERNELS}",
 )
-@pytest.mark.parametrize("cpu_template", SNAPSHOT_RESTORE_SHARED_NAMES["cpu_templates"])
 @pytest.mark.timeout(900)
 @pytest.mark.nonci
-def test_cpu_wrmsr_snapshot(bin_cloner_path, cpu_template):
+def test_cpu_wrmsr_snapshot(bin_cloner_path, msr_cpu_template):
```
Comment (pb8o) [inline tests/integration_tests/functional/test_cpu_features.py:466]: There doesn't seem to be any users of the `cpu_template` fixture ... Does it make sense to have two? or should we consolidate them into one?

File: resources/tests/msr/wrmsr_list.txt (hunk 3 above)
Comment (zulinx86) [inline resources/tests/msr/wrmsr_list.txt:17]: Can I ask why writing to MSR index 0x277 results in a vm crash?
And how did we choose the list of MSRs to be modified?

File: tests/integration_tests/functional/test_cpu_features.py (hunk 4 above)
Comment (sudanl0) [reply tests/integration_tests/functional/test_cpu_features.py:466]: Not all templates supported by test_feat_parity.py (C3, T2, T2CL, T2S, T2A) are supported by test_cpu_features (T2A, T2CL,T2S) so, we cannot use cpu_template fixture for rdmsr/wrmsr_cpuid snapshot/restore tests.
msr_cpu_template fixture was created local to test_cpu_features.py just for this reason.

File: tests/integration_tests/functional/test_cpu_features.py [hunk 5]
```diff
@@ -642,21 +639,13 @@ def _test_cpu_wrmsr_restore(context):
     )
 
 
-@pytest.mark.skipif(
-    PLATFORM != "x86_64", reason="CPU features are masked only on x86_64."
-)
```
[the remark this answers is not in the document]
Comment (sudanl0) [reply tests/integration_tests/functional/test_cpu_features.py:647]: SUPPORTED_CPU_TEMPLATES used here : https://github.com/firecracker-microvm/firecracker/blob/main/tests/integration_tests/functional/test_cpu_features.py#L225 
makes sure that we use the right template depending on the architecture. SUPPORTED_CPU_TEMPLATES https://github.com/firecracker-microvm/firecracker/blob/main/tests/framework/utils_cpu_templates.py#L20 internally checks if CpuVendor is either INTEL or AMD and returns templates based on that so it is safe to skip this step.

File: resources/tests/msr/wrmsr_list.txt (hunk 3 above)
Comment (sudanl0) [reply resources/tests/msr/wrmsr_list.txt:17]: AFAIK the wrmsr_list.txt was created by flipping 1 bit of the default value.
There could be MSRs which need to be update together and this could be the case here.
But to find a working set of MSRs to update we would have to go through reference doc which doesn't seem to be in the scope of the current task. Since the purpose of this file is to have some working combination of MSRs that we can update before running the snapshot test and also to have the file common for all Intel and AMD templates we decided to remove this 1 line.
As you see we don't update all the MSRs in this file for the same reason i.e. they are either reserved or need to be updated in a specific sequence.
Like in this case 0x277 is Page attribute table, its default value is 0x7040600070406 and as per AMD ref doc 3 is reserved value for bits 2:0.

File: tests/integration_tests/functional/test_cpu_features.py (hunk 4 above)
Comment (pb8o) [reply tests/integration_tests/functional/test_cpu_features.py:466]: Sure but is cpu_template used anywhere else now?

**Review by zulinx86 — APPROVED** [review]

**Review by pb8o — APPROVED** [review]

**Review by kalyazin — APPROVED** [review]

File: tests/integration_tests/functional/test_cpu_features.py (hunk 4 above)
Comment (sudanl0) [reply tests/integration_tests/functional/test_cpu_features.py:466]: test_feat_parity.py is the only one that is using it right now
